src/charm.py

```diff
@@ -1,104 +1,392 @@
+            logger.info(
+                "DU F1 port information not available. Using default value %s", DU_F1_DEFAULT_PORT
+            )
+            du_f1_port = DU_F1_DEFAULT_PORT
+        return _render_config_file(
+            gnb_name=self._gnb_name,
+            cu_f1_interface_name=self._charm_config.f1_interface_name,
+            cu_f1_ip_address=_get_pod_ip(),  # type: ignore[arg-type]
+            cu_f1_port=self._charm_config.f1_port,
+            du_f1_port=du_f1_port,
+            cu_n2_interface_name=self._charm_config.n2_interface_name,
+            cu_n2_ip_address=_get_pod_ip(),  # type: ignore[arg-type]
+            cu_n3_interface_name=self._charm_config.n3_interface_name,
+            cu_n3_ip_address=_get_pod_ip(),  # type: ignore[arg-type]
+            amf_external_address=self._n2_requirer.amf_ip_address,
+            mcc=self._charm_config.mcc,
+            mnc=self._charm_config.mnc,
+            sst=self._charm_config.sst,
+            tac=self._charm_config.tac,
+        )
+
+    def _is_cu_config_up_to_date(self, content: str) -> bool:
+        """Decide whether config update is required by checking existence and config content.
+
+        Args:
+            content (str): desired config file content
+
+        Returns:
+            True if config update is required else False
+        """
+        if not self._config_file_is_written() or not self._config_file_content_matches(
+            content=content
+        ):
+            return True
+        return False
+
+    def _config_file_is_written(self) -> bool:
+        return bool(self._container.exists(f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}"))
+
+    def _config_file_content_matches(self, content: str) -> bool:
+        if not self._container.exists(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}"):
+            return False
+        existing_content = self._container.pull(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}")
+        if existing_content.read() != content:
+            return False
+        return True
+
+    def _write_config_file(self, content: str) -> None:
+        self._container.push(source=content, path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}")
+        logger.info("Config file written")
+
+    def _configure_pebble(self, restart=False) -> None:
+        """Configure the Pebble layer.
+
+        Args:
+            restart (bool): Whether to restart the CU container.
+        """
+        plan = self._container.get_plan()
+        if plan.services != self._cu_pebble_layer.services:
+            self._container.add_layer(self._container_name, self._cu_pebble_layer, combine=True)
+            self._container.replan()
+            logger.info("New layer added: %s", self._cu_pebble_layer)
+        if restart:
+            self._container.restart(self._service_name)
+            logger.info("Restarted container %s", self._service_name)
+            return
+        self._container.replan()
```

Comment:
is this line necessary?

Reply:
No, I must have copied it accidentally from the above :) 
Removed.